tools/ci_build/build.py

```diff
@@ -1214,20 +1220,15 @@ def run_onnxruntime_tests(args, source_dir, ctest_path, build_dir, configs):
                 [sys.executable, 'onnxruntime_test_python.py'],
                 cwd=cwd, dll_path=dll_path)
 
-            iobinding_test = False
             # For CUDA enabled builds test IOBinding feature
-            if args.use_cuda:
-                iobinding_test = True
-                try:
-                    import torch  # noqa
-                except ImportError as error:
-                    iobinding_test = False
-                    log.exception(error)
-                    log.warning(
-                        "Torch is not installed. "
-                        "The IOBinding tests will be skipped as it requires Torch.")
-
-            if iobinding_test:
+            # Limit testing to Windows/Ubuntu and non-ARM builds as Torch wheels 
+            # may not be available for all platform/processor combinations
+            if args.use_cuda and (is_windows() or is_ubuntu_1604()) and not (args.arm or args.arm64):
```

Comment:
SHould cover the feature atleast on Windows in the CI pipeline

Reply:
so if i'm understanding correctly, it was previously failing on windows because torch was only installed for Linux (docker)
so now you are installing on both windows and linux?
wouldn't it have been better to keep your previous change (where it installs on Linux for both training and inference on gpu) , and install only windows here?
the installation would have been cached in most cases. 
now we will install on Linux gpu every single time?

Reply:
Not exactly.  The problem was having ` ${PYTHON_EXE} -m pip install --upgrade --pre torch==1.6.0.dev20200610 torchvision==0.7.0.dev20200610...` line for both training and inferencing CIs (basically put it outside the if). I was under the impression that this script ran in the Python packaging pipeline (and will thus keep IOBinding feature covered with tests atleast on Linux GPU builds). But, it turns out this script doesn't run there. And furthermore, `install_deps.sh` seems to be called from a variety of environments. One such is the Nuget GPU package tests in the Nuget GPU pipeline where it is called from a python 3.5 environment for which there are no Torch nightly dev wheels hosted - so a double whammy - neither was IOBinding tests beins run in the Python pipeline nor was the change to try and enable that keeping other pipelines happy.

So ended up reverting the change in `install_deps.sh` and installing Torch prior to running the IOBinding tests directly in build.py

Reply:
This change doesn't install it on Linux. The tests are covered only on Windows. Because, our Linux CI agents are not Ubuntu, I believe. I have a condition to check if the linux flavor is actually Ubuntu.THe reason why I am having this check is so that we don't run into a case where on some Linux flavor, it tries to install the Torch wheels only to not find them hosted and fail.

Reply:
I am wondering if we could limit this installation only when python bindings are enabled in the build.

---
In reply to: [452408457](https://github.com/microsoft/onnxruntime/pull/4462#discussion_r452408457) [](ancestors = 452408457)

Reply:
we are already doing that aren't we ? This whole section is under `if args.enable_pybind`